lib/services/local/headlessauthn_watcher_test.go

```diff
@@ -26,171 +26,324 @@ import (
+		buf.SetInit()
 
-	pubUUID := services.NewHeadlessAuthenticationID([]byte(sshPubKey))
+		watcherCtx, watcherCancel := context.WithCancel(ctx)
+		t.Cleanup(func() {
+			watcherCancel()
+		})
+
+		watcherClock := clockwork.NewFakeClock()
+		w, err := local.NewHeadlessAuthenticationWatcher(watcherCtx, local.HeadlessAuthenticationWatcherConfig{
+			Clock:          watcherClock,
+			WatcherService: buf,
+			Backend:        identity.Backend,
+		})
+		require.NoError(t, err)
 
-	waitInGoroutine := func(ctx context.Context, t *testing.T, name string, cond func(*types.HeadlessAuthentication) (bool, error)) (chan *types.HeadlessAuthentication, chan error) {
+		return &testSuite{
+			watcher:       w,
+			watcherClock:  watcherClock,
+			watcherCancel: watcherCancel,
+			identity:      identity,
+			buf:           buf,
+		}
+	}
+
+	waitInGoroutine := func(ctx context.Context, t *testing.T, watcher *local.HeadlessAuthenticationWatcher, name string, cond func(*types.HeadlessAuthentication) (bool, error)) (chan *types.HeadlessAuthentication, chan error) {
 		headlessAuthnCh := make(chan *types.HeadlessAuthentication, 1)
 		errC := make(chan error, 1)
 		go func() {
-			headlessAuthn, err := w.Wait(ctx, name, cond)
+			headlessAuthn, err := watcher.Wait(ctx, name, cond)
 			errC <- err
 			headlessAuthnCh <- headlessAuthn
 		}()
-		require.Eventually(t, func() bool { return w.CheckWaiter(name) }, time.Millisecond*100, time.Millisecond*10)
-
 		return headlessAuthnCh, errC
 	}
 
-	t.Run("WaitTimeout", func(t *testing.T) {
-		waitCtx, waitCancel := context.WithTimeout(ctx, time.Millisecond*10)
+	t.Run("WaitEventWithConditionMet", func(t *testing.T) {
+		t.Parallel()
+		s := newSuite(t)
+
+		waitCtx, waitCancel := context.WithTimeout(ctx, time.Second*5)
 		defer waitCancel()
 
-		_, err = w.Wait(waitCtx, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) { return true, nil })
-		require.Error(t, err)
-		require.Equal(t, waitCtx.Err(), err)
+		firstEmitReceived := make(chan struct{})
+		headlessAuthnCh, errC := waitInGoroutine(waitCtx, t, s.watcher, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
+			select {
+			case <-firstEmitReceived:
+			default:
+				close(firstEmitReceived)
+			}
+
+			return ha.User != "", nil
+		})
+
+		// Emit put event that doesn't pass the condition (user not set). The waiter should ignore these events.
+		stub, err := types.NewHeadlessAuthenticationStub(pubUUID, time.Now())
+		require.NoError(t, err)
+		stub.User = "user"
+		item, err := local.MarshalHeadlessAuthenticationToItem(stub)
+		require.NoError(t, err)
+
+		// The waiter may miss put events during initialization, so we continuously emit them until one is caught.
+		require.Eventually(t, func() bool {
+			s.buf.Emit(backend.Event{
+				Type: types.OpPut,
+				Item: *item,
+			})
+			select {
+			case <-firstEmitReceived:
+				return true
+			default:
+				return false
+			}
+		}, time.Second, time.Millisecond*100)
```

Comment:
TBH I'd increase the timeout to 2-3 seconds. When you run tests in parallel then it makes take longer to execute.